File: server/routes/Worker_auth.js
const express = require("express");
const Worker = require("../models/worker");
const Comment= require("../models/comment");

const bcryptjs = require('bcryptjs');
const jwt = require ("jsonwebtoken");
const auth = require("../middlewares/auth");
const User = require("../models/user");

const WorkerauthRouter = express.Router();

WorkerauthRouter.post("/api/Workersignup", async (req, res) =>{
    const {email, name , password , phone , type}= req.body;
    try{
        const existingUser = await Worker.findOne({email});

        if(existingUser){
            return res.status(400).json({ msg:"User with same email already exists"});
        }

        const hashedPassword = await bcryptjs.hash(password, 8);

        let worker = new Worker({
            email,
            name,
            password : hashedPassword,
            phone,
            type,
        })
       worker= await worker.save();
        res.json(worker);
    } catch(e) {
        res.status(500).json({error: e.message});
    }

    
    });

    WorkerauthRouter.post("/api/Workersignin", async (req, res) => {
        try{
            const{email , password } = req.body;

            const worker = await Worker.findOne({ email });

            if(!worker){
                return res.status(400).json({msg : "User with this email does not exit !"});
            }

            const isMatch = await bcryptjs.compare(password, worker.password);

            if(!isMatch){
                return res.status(400).json({msg : "Incorrect password!"});
            }

            const token = jwt.sign({id: worker._id}, "passwordKey");

            res.json({token, ...worker._doc})

        } catch(e) {
            res.status(500).json({ error: e.message});
        }
    });

    //get all business
    WorkerauthRouter.get("/business/get-all", async(req, res) => {
      try{
      const worker = await Worker.find();
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;
    
///get all halls
    WorkerauthRouter.get("/business/get-halls", async(req, res) => {
        try{
        const worker = await Worker.find({type: 'halls'});
        res.json(worker);
        }catch (e) {
          res.status(500).json({ error: e.message});
        }
      }) ;
      

      ///get halls by the nablus 
    WorkerauthRouter.get("/business/get-Nablus", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'halls',location:'Nablus'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;
    

    ///get halls by the nablus 
    WorkerauthRouter.get("/business/get-Tulkarem", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'halls',location:'Tulkarem'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;

    ///get halls by the Ramallah

    WorkerauthRouter.get("/business/get-Ramallah", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'halls',location:'Ramallah'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;

    ///get halls by the nablus 
    WorkerauthRouter.get("/business/get-Nablusband", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'Band group',location:'Nablus'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;
    

    ///get band by the nablus 
    WorkerauthRouter.get("/business/get-Tulkaremband", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'Band group',location:'Tulkarem'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;

    ///get halls by the Ramallah

    WorkerauthRouter.get("/business/get-Ramallahband", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'Band group',location:'Ramallah'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;


      ///get all band group
    WorkerauthRouter.get("/business/get-band", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'Band group'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;
    /// get all photography 
    WorkerauthRouter.get("/business/get-photography", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'photography'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;

    //get all decorate

    WorkerauthRouter.get("/business/get-decorate", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'Decorate'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;
    /// get all photography 
    WorkerauthRouter.get("/business/get-photography", async(req, res) => {
      try{
      const worker = await Worker.find({type: 'photography'});
      res.json(worker);
      }catch (e) {
        res.status(500).json({ error: e.message});
      }
    }) ;



      WorkerauthRouter.post("/api/rate-Worker", auth, async (req, res) => {
        try {
          const { id, rating } = req.body;
          let worker = await Worker.findById(id);
      
          for (let i = 0; i < worker.ratings.length; i++) {
            if (worker.ratings[i].userId == req.user) {
              worker.ratings.splice(i, 1);
              break;
            }
          }
      
          const ratingSchema = {
            userId: req.user,
            rating,
          };
      
          worker.ratings.push(ratingSchema);
          worker = await worker.save();
          res.json(worker);
        } catch (e) {
          res.status(500).json({ error: e.message });
        }
      });

      WorkerauthRouter.post("/api/comment-Worker", auth, async (req, res) => {
        try {
          const { id, comment ,username } = req.body;
          let worker = await Worker.findById(id);
      
          for (let i = 0; i < worker.comments.length; i++) {
            if (worker.comments[i].userId == req.user) {
              worker.comments.splice(i, 1);
              break;
            }
          }
      
          const commentSchema = {
            userId: req.user,
            comment,
            username: username,
          };
      
          worker.comments.push(commentSchema);
          worker = await worker.save();
          res.json(worker);
        } catch (e) {
          res.status(500).json({ error: e.message });
        }
      });

      WorkerauthRouter.post("/api/order-Worker", auth, async (req, res) => {
        try {
          const { eventname,id, date ,from,to,username,userphone,status } = req.body;
          let worker = await Worker.findById(id);
      
          for (let i = 0; i < worker.orders.length; i++) {
            if (worker.orders[i].userId == req.user) {
              worker.orders.splice(i, 1);
              break;
            }
          }
      
          const orderSchema = {
            userId: req.user,
            eventname,
            date,
            from,
            to,
            username: username,
            userphone:userphone,
            status
          };
      
          worker.orders.push(orderSchema);
          worker = await worker.save();
          res.json(worker);
        } catch (e) {
          res.status(500).json({ error: e.message });
        }
      });

      

      WorkerauthRouter.get("/business/search/:name", auth, async (req, res) => {
        try {
          const worker = await Worker.find({
            name: { $regex: req.params.name, $options:"i"},
          });
         
          res.json(worker);
        } catch (e) {
          res.status(500).json({ error: e.message });
        }
      });

      //adminsearch
      WorkerauthRouter.get("/business/adminsearch/:name", auth, async (req, res) => {
        try {
          const worker = await Worker.find({
            name: { $regex: req.params.name, $options:"i"},
          });
         
          res.json(worker);
        } catch (e) {
          res.status(500).json({ error: e.message });
        }
      });

      /////show all comments

      WorkerauthRouter.get("/business/searchh/:name", async (req, res) => {
         try {
          const worker = await Worker.find({
            name: req.params.name,
          });
        
          res.json(worker);
        } catch (e) {
          res.status(500).json({ error: e.message });
        }
      });

      // Delete the product
      WorkerauthRouter.post("/admin/delete-product", async (req, res) => {
  try {
    const { id } = req.body;
    let worker = await Worker.findByIdAndDelete(id);
    res.json(worker);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});
//Delete users 
WorkerauthRouter.post("/admin/delete-user", async (req, res) => {
  try {
    const { id } = req.body;
    let user = await User.findByIdAndDelete(id);
    res.json(user);
  } catch (e) {
    res.status(500).json({ error: e.message });
  }
});

///get all workers
WorkerauthRouter.get("/business/all", async(req, res) => {
  try{
  const worker = await Worker.find({statu:'not_activate'});
  res.json(worker);
  }catch (e) {
    res.status(500).json({ error: e.message});
  }
}) ;

//update worker info
WorkerauthRouter.post("/api/update-info", async (req, res) =>{
  const { id ,name , email , phone}= req.body;
  let worker = await Worker.findOne({id})
      
      worker.name=name;
      worker.email=email;
      worker.phone=phone;
       worker = await worker.save();
       res.json(worker);

  try{

  }catch(e) {
      res.status(500).json({ error: e.message });
    }
});




module.exports = WorkerauthRouter;

File: lib/services/workerauth.dart
import 'dart:convert';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_try/constants/error_handling.dart';
import 'package:flutter_try/constants/global_variables.dart';
import 'package:flutter_try/constants/utils.dart';
import 'package:flutter_try/models/user.dart';
import 'package:flutter_try/models/worker.dart';
import 'package:flutter_try/providers/user_provider.dart';
import 'package:flutter_try/providers/worker_provider.dart';
import 'package:http/http.dart' as http;
import 'package:provider/provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

class WorkerAuthService {

  //signup user
  void signUpUser({
    required BuildContext context,
    required String email,
    required String name,
    required String password,
    required String phone,
    required String type,
   // required String discreption,
    //required List<String> images,
    //required double price,
    //required String location,

  }) async{
    try{
      Worker worker = Worker(id: '', 
      email: email,
       name: name,
        password: password,
         phone: phone,
         type : type,
         discreption: '',
         images: [],
         price: 0.0,
         location: '',
         latitude: 0.0,
         longitude: 0.0,
         statu: '',
          token: '',

          );

          http.Response res = await http.post(
            Uri.parse('$uri/api/Workersignup'),
            body: worker.toJson(),
            headers: <String, String>{
            'content-Type': 'application/json; charest=UTF-8',
            },
            );
            httpErrorHandel(response: res,
             context: context,
              onSuccess: (){
                showSnackBar(context, 'Account created ! Login with the same created',);
              },
              );


    } catch(e) {
      showSnackBar(context, e.toString());

    }

  }

  // sign in user
  void signInUser({
    required BuildContext context,
    required String email,
    required String password,
  }) async {
    try {
      http.Response res = await http.post(
        Uri.parse('$uri/api/Workersignin'),
        body: jsonEncode({
          'email': email,
          'password': password,
        }),
        headers: <String, String>{
          'Content-Type': 'application/json; charset=UTF-8',
        },
      );
      httpErrorHandel(
        response: res,
        context: context,
        onSuccess: () async {
          SharedPreferences prefs = await SharedPreferences.getInstance();
          Provider.of<WorkerProvider>(context, listen: false).setWorker(res.body);
          await prefs.setString('x-auth-token', jsonDecode(res.body)['token']);
          Navigator.of(context).pushNamed('businessScreen');Colors.red[200];

          
        },
      );
    } catch (e) {
      showSnackBar(context, e.toString());
    }
  }

  //get all business

  Future<List<Worker>>showAll(BuildContext context) async {
  
  
  List <Worker> alls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-all'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            alls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return alls;

}

  ////get all halls

  Future<List<Worker>>showAllhalls(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-halls'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

Future<List<Worker>>showhallsNablus(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-Nablus'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

///show Ramallah halls
Future<List<Worker>>showhallsRamallah(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-Ramallah'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

Future<List<Worker>>showhallsTulkarem(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-Tulkarem'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

//get band Nablud,Ranmallah tulkarem

Future<List<Worker>>showbandNablus(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-Nablusband'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

///show Ramallah halls
Future<List<Worker>>showbandRamallah(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-Ramallahband'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

Future<List<Worker>>showbandTulkarem(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-Tulkaremband'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}


Future<List<Worker>>showrecomendedhalls(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-halls'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

////get all band group

  Future<List<Worker>>showAllband(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-band'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

Future<List<Worker>>showrecomendedband(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-band'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}
//get all photography
Future<List<Worker>>showAllphotograpth(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-photography'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}
//get all decorate

Future<List<Worker>>showAlldecorate(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/get-decorate'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}


void deleteProduct({
    required BuildContext context,
    required Worker worker,
    required VoidCallback onSuccess,
  }) async {
    final userProvider = Provider.of<UserProvider>(context, listen: false);

    try {
      http.Response res = await http.post(
        Uri.parse('$uri/admin/delete-product'),
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
          'x-auth-token': userProvider.user.token,
        },
        body: jsonEncode({
          'id': worker.id,
        }),
      );

      httpErrorHandel(
        response: res,
        context: context,
        onSuccess: () {
          onSuccess();
        },
      );
    } catch (e) {
      showSnackBar(context, e.toString());
    }
  }

  ////get all workers

  Future<List<Worker>>allworkers(BuildContext context) async {
  
  
  List <Worker> workerhalls =[];
  try {
          http.Response res = await http.get(Uri.parse('$uri/business/all'), headers:{
          'Content-Type': 'application/json; charset=UTF-8',
          }
          ) ;   

          httpErrorHandel(response: res, context: context, onSuccess: () {
            for (int i =0; i < jsonDecode(res.body).length;i++){
            workerhalls.add(Worker.fromJson(jsonEncode(jsonDecode(res.body)[i],
              ),
              ),
              );
            }
          });
  } catch (e) {
              showSnackBar(context, e.toString());
  }
  return workerhalls;

}

//update business info

Future <Worker>  updatinfo({
required BuildContext context,
required String id,
required String name,
required String email,
required String phone,





}) async {
  final workerProvider = Provider.of<WorkerProvider>(context, listen: false);
   late Worker wo = workerProvider.worker ;
  try {
       
       http.Response res = await http.post(
        Uri.parse('$uri/api/update-info'),
        
        headers: {
          'Content-Type': 'application/json; charset=UTF-8',
          'x-auth-token': workerProvider.worker.token,
        },
        body: jsonEncode({
          'name':name,
          'email':email,
          'phone':phone,
        }), );
        httpErrorHandel(
        response: res,
        context: context,
        onSuccess: () {
          wo=jsonDecode(res.body);
          showSnackBar(context, 'update succefully',);
         // Navigator.pop(context);
        },
      ); 
      
      


       

  }catch(e){
    showSnackBar(context, e.toString());
  }
  return wo;

}

  

}